be/src/storage/rowset/column_writer.cpp

```diff
@@ -816,4 +871,136 @@ Status StringColumnWriter::check_string_lengths(const Column& column) {
     return Status::OK();
 }
 
+DictColumnWriter::DictColumnWriter(const ColumnWriterOptions& opts, TypeInfoPtr type_info,
+                                   std::unique_ptr<ScalarColumnWriter> column_writer)
+        : ColumnWriter(std::move(type_info), opts.meta->length(), opts.meta->is_nullable()),
+          _scalar_column_writer(std::move(column_writer)) {}
+
+Status DictColumnWriter::append(const Column& column) {
+    if (_is_speculated) {
+        return _scalar_column_writer->append(column);
+    }
+
+    if (_buf_column == nullptr) {
+        // first column size is greater than speculate size
+        if (column.size() >= config::dictionary_speculate_min_chunk_size) {
+            _is_speculated = true;
+            RETURN_IF_ERROR(speculate_column_and_set_encoding(column));
+            return _scalar_column_writer->append(column);
+        } else {
+            _buf_column = column.clone_empty();
+            _buf_column->append(column, 0, column.size());
+            return Status::OK();
+        }
+    }
+    _buf_column->append(column, 0, column.size());
```

Comment:
We should check if _buf_column overflow before append.

Reply:
check this bugfix https://github.com/StarRocks/starrocks/pull/36281

Reply:
@luohaha updated